### DISCUSSION TITLE
Update from SD card

### DISCUSSION THREAD

**Discussion by saschagu** [biologist79/ESPuino#63] [Ideas]
Hi all,
I added code from the OTA update example ([1]) and could update the firmware from SD card via FTP. I used min_spiffs.csv as partition layout.
Is there a particular reason you switched to hugh_app.csv as default partition layout? This makes OTA update impossible. With min_spiffs I still have > 400kb Flash memory left, even with MQTT, FTP etc. activated...
The OTA update feature could be interesting for people that don't have easy access to their ESP32 (like me) and still want to update to new features.

[1] https://github.com/espressif/arduino-esp32/blob/master/libraries/Update/examples/SD_Update/SD_Update.ino

**Comment by biologist79** [comment]
The reason is: I did some tests with bluetooth (a2dp-sink). Depending on the modules (MQTT, FTP...) choosed, flash-memory is the limiting-factor. BT is currently not (really) implemented but I perspectively changed the partition-layout to huge_app.

So for now you don't "miss" anything by reverting it back to min_spiffs. Mid-term I guess there'll be some free flash-memory by moving html-stuff from flash to SD. Especially for i18n this will be necessary. What I don't like about this is solution is, that the user has additionaly to take care that html-files on SD are up to date as well. That's why I not already did it.

**Reply by saschagu** [reply]
I agree, HTML on the SD card might be confusing and error-prone too. Do you think its worth integrating the OTA code at least with a configuration option? I could start a pull request...

**Reply by biologist79** [reply]
To be honest, implementing OTA was one of my initial goals. In the meantime this didn't work because of static memory-issues (took me a lot of time to point out the reason), so I didn't try anymore (albeit having that point still in my mind). However, good to see it's indeed working now.

My suggestion is: what a bit with the PR until I re-arranged the code.

**Reply by tuniii** [reply]
I am contributing to an open source project which also uses the ESP32. I have some experience at OTA and building firmware automated via Github Actions for the ESP32. I discovered this project recently and I want to build some ESPuinos. It would be great to have the OTA feature and some prebuilt firmwares in future. Are there any plans to restructure the code to continue with the OTA implementation? I could also help with the OTA implementation and the automated firmware building.

**Reply by biologist79** [reply]
Indeed one of my initial goals was to provide OTA-feature. Meanwhile I had some problems (as it turned out because of lack of memory) and after that I never tried again. But I've been told by others it's working - so in general it seems to work. Having said this: yes, would be fine to support OTA.
But to be honest I don't know how to provide automatic firmwares because there's tons of options that can be configured via settings.h for instance. How do you want to address this?

**Reply by tuniii** [reply]
There are basically two ways to address this. Some of the options can be moved out of the preprocessor to make them configurable via the web interface, e.g. pin configuration, sd card mode. For the options with a big memory foot print, several firmwares can be generated by passing the preprocessor flags via the console during building.

**Reply by tuniii** [reply]
I created a fork to demonstrate [Github Actions](https://github.com/tuniii/ESPuino/actions). Have a look at the latest build and its artifacts.
Check also [Releases](https://github.com/WLANThermo-nano/WLANThermo_ESP32_Software/releases) of another project as a reference how releases can be build for different hardwares.

**Reply by biologist79** [reply]
Interesting :-)
However, would need intensively testing. To be honest: don't have time currently to do that. So I don't say NO but don't expect this to be happen short-term. In general this idea isn't new, but having firmwares built automatically could help.

SD is a bit more complicated as it's mandatory for ESPuino to run. However, it could be moved to the "simple" GUI that is used to configure WiFi-credentials+hostname. Hmm, will test - could be a good point to start with.

**Comment by biologist79** [comment]
Recently I enabled OTA for boards with 16 MB flash-memory. Update can be performed via webGUI.
